### ISSUE TITLE
Taking MDS matrices as from [DL18] or their transposes?

### ISSUE THREAD

**Issue by weikengchen** [vesselinux/anemoi-hash#9]
In the Sage script, let us assume that we use generator g = 7 (for BLS12-381 scalar field), then the corresponding MDS matrix (retrieved from `A.mat`) is as follows:

| | col-0 | col-1 | col-2 |
| -------------| ------------- | ------------- | ------------- |
| **row-0** | 8 | 1 |  7 |
| **row-1** | 1 | 1 | 1 |
| **row-2** | 8 | 7 | 1 |

So, given x = [1, 2, 3], the result is [31, 6, 25].

However.

This is somewhat different from what we can find on page 14, and also different from how Figure 6 in [DL18] suggests (note, [DL18] does not just provide a matrix, but also a diagram that describes this map). In these versions, the result should be [34, 24, 12].

It is not wrong for the Sage implementation to use the transpose of these MDS matrices. By definitions, MDS matrices observe the minors of the matrices, and the transpose has the same minors here. That is, the matrix above (which is the transpose of the one in [DL18]) is a good MDS matrix.

Therefore, the main reason for this issue is to ask if the Sage implementation wants to be made consistent with the paper so that we can avoid conflicting implementations in the future---this is to establish reference implementations as well as test vectors that can be later standardized.

**Comment by weikengchen** [comment]
Another related issue is that, for 4x4 MDS matrix, the `A.mat` obtained from:

```
A = AnemoiPermutation(q=0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001, 
                      alpha=5, n_rounds=11, n_cols=4)
```

Its **transpose** is as follows.

| | col-0 | col-1 | col-2 | col-3 |
| -------------| ------------- | ------------- | ------------- | ------------- |
| **row-0** | 1 | 49 | 49 | 8 |
| **row-1** | 8 | 56 | 49 | 15 |
| **row-2** | 7 | 8 | 1 | 7 |
| **row-3** | 7 | 15 | 8 | 8 |

This matrix has some similarity with the one in the paper---its transpose, with some shifting, only differs from the one in paper in that two of "1" becomes "15" (row-1 and col-3, row-3 and col-1).

**Comment by Nashtare** [comment]
You're right that the SAGE script doesn't take the _actual_ matrices described in DL18 and should be adapted accordingly. Note however that the specs are still under discussion, and for performance reasons, we may switch to different matrices.
We'll update the eprint version accordingly once we settle on this (with decision reflected on the SAGE script).

**Comment by Nashtare** [comment]
Hey @weikengchen, we've updated the paper and adjusted the matrices for `\ell = 3, 4`. We've also adjusted the generation in the SAGE script here (and updated the documentation) so that the fast algorithms presented as `M_3` and `M_4` correspond to the fast matrix-vector multiplication happening in the MDS layer.
All the instantiations presented here (for 128bits security) have been implemented [here](https://github.com/Nashtare/anemoi) following the specs.
